push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,122 @@
+
+## OAuth provider support 
+
+To address the first & second problem, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
+The OAuth flow will follow the same pattern & flow used by the [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy).
+The UPS APB will need to be modified to include an OAuth ServiceAccount, similar to how [Jenkins uses a ServiceAccount](https://github.com/openshift/jenkins-openshift-login-plugin#browser-access).
+UPS will use this ServiceAccount as an OAuth Client to verify the User has the necessary permissions in OpenShift to use UPS.
+The permissions will be:
+* edit access on the namespace in OpenShift => Standard Developer Account in UPS
+* admin access on the namespace in OpenShift => Admin Account in UPS
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Minimal changes will be required to the UPS APB.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift Integration
+
+To address the third problem, a sidecar will be added to the UPS APB.
+This sidecar will provide background bi-directional syncing of resources and metadata between OpenShift and UPS.
+The sidecar will be written in Go and utilise a kubernetes client for watching & updating Kubernetes resources.
+It will use a SerivceAccount for auth when watching resources.
+The sidecar will also use a ServiceAccount to access the UPS Admin REST API for reading & updating UPS resources.
+
+### Syncing
+
+Syncing will happen *from* OpenShift *to* UPS for the following resources:
+
+* MobileClient => Variant
+* Secret (p12 or secret key) => Variant configuration
+
+The Secrets will have an appropriate label so it can be easily filtered when looking for resources to sync.
+The sidecar will overwrite the Variant fields and configuration to match what the values are in OpenShift.
+This means that any changes to a Variant directly in the UPS UI will be overwritten on next sync.
+
+Syncing will happen *from* UPS *to* OpenShift for the following resources:
+
+* Variant push configuration (JSON) => MobileClient annotation
+
+In UPS, a JSON object with Push configuration for a Variant is available.
+However, this JSON object isn't part of the generated mobile-services.json file from the Mobile CLI or UI.
+This JSON config will be read from UPS and synced back as an annotation on the associated MobileClient in OpenShift.
+Any manual changes this annotation will be overwritten by the sidecar when it next syncs.
+
+```json
+{
+  // TODO: example mobile-services.json structure with a single variant
```

Comment:
Syncing with @wtrocki about push, we were thinking if we really need to provide this one. Today the developers need to take care of it yourself, they need to go the UI and copy and paste the JSON provided by the UPS. It also kills the problem with more than one variant.

Reply:
@danielpassos @wtrocki Do you see value in any element of syncing?

Do you think the goal should be to have a single config file (mobile-services.json) with config for all services, including UPS?

Reply:
I think we can have push included in mobile-services.json, but I don't know if we should provide the variants information there, or if so, how to provide this information there.

I think we can have at least the push URL or so only to _flag_ there is a push services there. I'm not 100% sure about that yet. Something like...

```json
{
    "id": "push",
    "name": "push",
    "url": "http://someurl/push",
    "type": "push",
    "config": {
    }
}
```

If we decide to provide the variants we need to think what is the best way to provide it (talking about apb provide it and SDK to comsume it.

```json
{
    "id": "push",
    "name": "push",
    "url": "http://someurl/push",
    "type": "push",
    "config": {
        "android": [
            {
                "name": "driver",
                "variant": "123456789-123456789",
                "secret": "987654321-987654321",
                "sender": "9999"
            },
            {
                "name": "passeger",
                "variant": "123456789-123456789",
                "secret": "987654321-987654321",
                "sender": "9999"
            }
        ],
        "ios": {
            "name": "admin",
            "variant": "123456789-123456789",
            "secret": "987654321-987654321"
        }
    }
}
```

But how to create/edit/remove the variants informations?

Reply:
+1 for only url in config. 
I do not see value in sync - it will introduce a lots of additional work for very little value. Users can provision apb and use the same configurations steps as upstream

Reply:
> But how to create/edit/remove the variants informations?

See latest comment with a suggested flow which shows were/when this could be created.

Reply:
I'd rather see that per MAR / backend serivce.

We should NEVER, EVER, expose the MasterSecret, which allows you to send push notifications through the API, on the device config - on the device, only _ITS OWN_ variant and variantSecret should be there

Reply:
@matzew Agree completely on never having the master secret on device (not sure where that was suggested).
The master secret could be used by a custom service though for dynamically sending push messages via the sender api